### PR TITLE
NETOBSERV-1689 Allow multi-tenant Prometheus queries

### PR DESCRIPTION
**Pull request by jpinsonneau**
## Description

- manage prom `devUrl`
- grab metrics containing `namespace` only
- skip default quick filters since not relevant for developers

image
image
image

To try this PR, you can create a user and assign it to a namespace:
```
oc create user test
oc adm policy add-role-to-user edit test -n openshift-console
```

Then impersonate the user from the console to access developer perspective.

## Dependencies

https://github.com/netobserv/network-observability-operator/pull/726

## Checklist

If you are not familiar with our processes or don't know what to answer in the list below, let us know in a comment: the maintainers will take care of that.

* [X] Is this PR backed with a JIRA ticket? If so, make sure it is written as a title prefix _(in general, PRs affecting the NetObserv/Network Observability product should be backed with a JIRA ticket - especially if they bring user facing changes)._
* [x] Does this PR require product documentation?
  * [ ] If so, make sure the JIRA epic is labelled with "documentation" and provides a description relevant for doc writers, such as use cases or scenarios. Any required step to activate or configure the feature should be documented there, such as new CRD knobs.
* [ ] Does this PR require a product release notes entry?
  * [ ] If so, fill in "Release Note Text" in the JIRA.
* [ ] Is there anything else the QE team should know before testing? E.g: configuration changes, environment setup, etc.
  * [ ] If so, make sure it is described in the JIRA ticket.
* QE requirements (check 1 from the list):
  * [X] Standard QE validation, with pre-merge tests unless stated otherwise.
  * [ ] Regression tests only (e.g. refactoring with no user-facing change).
  * [ ] No QE (e.g. trivial change with high reviewer's confidence, or per agreement with the QE team).

### CODE REVIEW

**Comment by jotak** [comment]
I'm testing and this is working well for prometheus queries, however this is breaking the Loki queries:

```
time="2024-08-02T10:58:37Z" level=debug msg="GetFlows query params: map[dataSource:[auto] dedup:[false] filters:[] limit:[50] namespace:[netobserv] packetLoss:[all] rateInterval:[30s] recordType:[flowLog] step:[15s] timeRange:[300] type:[Flows]]" module=handler
time="2024-08-02T10:58:37Z" level=debug msg="executeLokiQuery URL: http://loki.netobserv.svc:3100/loki/api/v1/query_range?query={app=\"netobserv-flowcollector\"}|~`namespace\":\"netobserv\"`&start=1722596017&limit=50" module=handler
```
As you can see, it injects a filter `namespace\":\"netobserv\"`, which isn't valid with Loki, it should be using SrcK8S_Namespace instead

**Comment by jpinsonneau** [comment]
> I'm testing and this is working well for prometheus queries, however this is breaking the Loki queries:
> 
> ```
> time="2024-08-02T10:58:37Z" level=debug msg="GetFlows query params: map[dataSource:[auto] dedup:[false] filters:[] limit:[50] namespace:[netobserv] packetLoss:[all] rateInterval:[30s] recordType:[flowLog] step:[15s] timeRange:[300] type:[Flows]]" module=handler
> time="2024-08-02T10:58:37Z" level=debug msg="executeLokiQuery URL: http://loki.netobserv.svc:3100/loki/api/v1/query_range?query={app=\"netobserv-flowcollector\"}|~`namespace\":\"netobserv\"`&start=1722596017&limit=50" module=handler
> ```
> 
> As you can see, it injects a filter `namespace\":\"netobserv\"`, which isn't valid with Loki, it should be using SrcK8S_Namespace instead

oh good catch; I forgot about that case and I admitt my tests were focusing on Prom.
Let me update the PR to solve this

**Comment by jpinsonneau** [comment]
added a skip for loki case + rebased

https://github.com/netobserv/network-observability-console-plugin/pull/571/commits/95620583110272ab64176e5a6c388c651f84685d

@jotak does that work for you to simply skip this filter as same as prom case ? 
The Loki gateway will show what's authorized. If the user have access to multiple namespaces; he will see everything at first.

**Comment by jotak** [comment]
@jpinsonneau hmm so that would ignore the global project selection for loki queries .. I don't think we should ignore it.
I have the feeling we should consider it not as a filter but as a special parameter in `FlowQueryBuilder`  (and maybe also in `TopologyInput` ?) - that indeed can result in creation of a filter, but we see that it follows specific rules - let me see if that would work better

**Comment by jpinsonneau** [comment]
> @jpinsonneau hmm so that would ignore the global project selection for loki queries .. I don't think we should ignore it. I have the feeling we should consider it not as a filter but as a special parameter in `FlowQueryBuilder` (and maybe also in `TopologyInput` ?) - that indeed can result in creation of a filter, but we see that it follows specific rules - let me see if that would work better

Sure @jotak. Feel free to add commits here if you find an interesting approach :wink: 
Thanks !

**Comment by jpinsonneau** [comment]
Merged @jotak changes here
https://github.com/netobserv/network-observability-console-plugin/pull/571/commits/5f5d4fbca6121ab30306dc0cdd75c9056c97e4ba

File: pkg/prometheus/query.go [hunk 1]
```diff
@@ -145,6 +145,10 @@ func appendFilteredMetric(sb *strings.Builder, metric string, filters filters.Si
 	sb.WriteRune('{')
 	first := true
 	for _, filter := range filters {
+		// namespace filtering is managed by prom label proxy so we can simply skip it
+		if filter.Key == "namespace" {
```
Comment (jotak) [inline pkg/prometheus/query.go:149]: I forgot to remove that in my last commit, it shouldn't be useful anymore

**Review by jotak — APPROVED** [review]
LGTM

@jpinsonneau I pushed a small commit to remove some leftovers I forgot to clean before

**Pull request merged**